### PR TITLE
Handle Azure long-running operations

### PR DESCRIPTION
**Pull request by brendandixon**
Signed-off-by: Brendan Dixon <brendand@microsoft.com>

### CODE REVIEW

File: autorest/azure/azure.go [hunk 1]
```diff
@@ -126,6 +131,58 @@ func ExtractRequestID(resp *http.Response) string {
...
 }
 
+// GetAsyncOperation retrieves the long-running URL to poll from the passed response.
+func GetAsyncOperation(resp *http.Response) string {
+	return resp.Header.Get(http.CanonicalHeaderKey(HeaderAsyncOperation))
+}
+
+// ResponseIsLongRunning returns true if the passed response is for an Azure long-running operation.
+func ResponseIsLongRunning(resp *http.Response) bool {
+	return autorest.ResponseRequiresPolling(resp, http.StatusCreated) && len(GetAsyncOperation(resp)) > 0
```
Comment (ahmetalpbalkan) [inline autorest/azure/azure.go]: I think we ought to handle both `http.StatusCreated` (201)  and `http.StatusAccepted` (202) for long-running operations.

`Azure-AsyncOperation` header is present in both response status codes (unless the operation is completed right away sometimes with HTTP 201, then we should not poll). 

Also let's use `== ""` comparison for strings.

File: autorest/azure/azure.go [hunk 2]
```diff
@@ -126,6 +131,58 @@ func ExtractRequestID(resp *http.Response) string {
...
+}
+
+// NewAsyncPollingRequest allocates and returns a new http.Request to poll an Azure long-running
+// operation. If it successfully creates the request, it will also close the body of the passed
+// response, otherwise the body remains open.
+func NewAsyncPollingRequest(resp *http.Response, c autorest.Client) (*http.Request, error) {
```
Comment (ahmetalpbalkan) [inline autorest/azure/azure.go]: Where is this going to get called (sdk?)

File: autorest/azure/azure.go (hunk 1 above)
Comment (brendandixon) [inline autorest/azure/azure.go]: 202 is handled by other code and, as per the current Azure REST specifications, relies on the Location header. If that's changed, then we can adapt this code.

File: autorest/azure/azure.go (hunk 2 above)
Comment (brendandixon) [inline autorest/azure/azure.go]: Not to worry. I sent mail outlining the full scope of changes to Garima.

File: autorest/azure/azure.go (hunk 1 above)
Comment (ahmetalpbalkan) [inline autorest/azure/azure.go]: @brendandixon hmm right. Although, in this case Azure APIs (specifically Storage RP) that return both `Location` and `Azure-AsyncOperation` header send different URLs in it. So there may be a behavioral difference. I'd say if the endpoint is specified as long running operation in Swagger, let's handle it in one place.

But before doing that, let's give it a try like this, I think even though these headers point to different URLs, they behaviorally do the same thing.

File: autorest/azure/azure.go (hunk 2 above)
Comment (ahmetalpbalkan) [inline autorest/azure/azure.go]: alright, I appreciate if you can fwd.

File: autorest/azure/azure.go (hunk 1 above)
Comment (brendandixon) [inline autorest/azure/azure.go]: We must follow the specification. The last time I read it, it outlined the implemented behavior. If that's changed, I want to see the update before I make code changes.

**Comment by ahmetalpbalkan** [comment]
LGTM feel free to merge.

**Pull request merged**